lib/modules.nix

```diff
@@ -230,12 +230,20 @@ rec {
      correspond to the definition of 'loc' in 'opt.file'. */
   mergeOptionDecls = loc: opts:
     foldl' (res: opt:
-      if opt.options ? default && res ? default ||
-         opt.options ? example && res ? example ||
-         opt.options ? description && res ? description ||
-         opt.options ? apply && res ? apply ||
-         # Accept to merge options which have identical types.
-         opt.options ? type && res ? type && opt.options.type.name != res.type.name
+      let t  = res.type;
+          t' = opt.options.type;
+          mergedType = t.typeMerge t'.functor;
+          typeSet = if (bothHave "type") && sameType && (mergedType != null)
+                       then { type = mergedType; }
+                       else {};
```

Comment:
If both inputs have a type, then the output should also have a type.  I suggest to remove the other parts of the condition.